railties/lib/rails/commands/server/server_command.rb

```diff
@@ -140,21 +80,72 @@ def restart_command
 
   module Command
     class ServerCommand < Base # :nodoc:
-      def help
-        puts Rails::Server::Options.new.option_parser(Hash.new)
+      DEFAULT_PID_PATH = File.expand_path("tmp/pids/server.pid").freeze
+
+      class_option :port, aliases: "-p", type: :numeric,
+        desc: "Runs Rails on the specified port.", banner: :port
+      class_option :binding, aliases: "-b", type: :string,
+        desc: "Binds Rails to the specified IP.", banner: :IP
+      class_option :config, aliases: "-c", type: :string, default: "config.ru",
+        desc: "Uses a custom rackup configuration.", banner: :file
+      class_option :daemon, aliases: "-d", type: :boolean, default: false,
+        desc: "Runs server as a Daemon."
+      class_option :environment, aliases: "-e", type: :string,
+        desc: "Specifies the environment to run this server under (test/development/production).", banner: :name
+      class_option :pid, aliases: "-P", type: :string, default: DEFAULT_PID_PATH,
+        desc: "Specifies the PID file."
+      class_option "dev-caching", aliases: "-C", type: :boolean, default: nil,
+        desc: "Specifies whether to perform caching in development."
+
+      def initialize(*)
+        super
+        @server = args.shift
+        @log_stdout = options[:daemon].blank? && (options[:environment] || Rails.env) == "development"
       end
 
       def perform
         set_application_directory!
-
-        Rails::Server.new.tap do |server|
+        Rails::Server.new(server_options).tap do |server|
           # Require application after server sets environment to propagate
           # the --environment option.
           require APP_PATH
           Dir.chdir(Rails.application.root)
           server.start
         end
       end
+
+      private
+        def server_options
+          {
+            server:       @server,
+            log_stdout:   @log_stdout,
+            Port:         port,
+            Host:         host,
+            config:       options[:config],
+            daemonize:    options[:daemon],
+            environment:  environment,
+            pid:          options[:pid],
+            caching:      options["dev-caching"],
+            AccessLog:    [],
+            restart_cmd:  restart_command
+          }
+        end
+
+        def port
+          options[:port] || ENV.fetch("PORT", 3000).to_i
+        end
+
+        def host
+          options[:binding] || ENV.fetch("HOST", "localhost")
+        end
+
+        def environment
+          options[:environment] || (ENV["RAILS_ENV"] || ENV["RACK_ENV"] || "development")
+        end
+
+        def restart_command
+          "bin/rails server #{ARGV.join(' ')}"
```

Comment:
I don't think the command infrastructure guarantees argv to be unharmed. We should add one test that uses every option and ensures every one option is present in the rerun command.